CMakeLists.txt

```
@@ -0,0 +1,164 @@
+# contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+# file except in compliance with the License. You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+# Description: Core CMake rules.
+
+cmake_minimum_required( VERSION 3.16.3 )
+project(
+    microlibrary
+    LANGUAGES CXX
+)
+
+# general configuration
+
+set(
+    MICROLIBRARY_HIL
+    "" CACHE STRING
+    "microlibrary: HIL."
+    )
+
+set(
+    MICROLIBRARY_TARGET
+    "" CACHE STRING
+    "microlibrary: Target environment."
+    )
+
+option(
+    MICROLIBRARY_BUILD_AUTOMATED_TESTS
+    "microlibrary: Build automated tests."
+    OFF
+    )
+
+option(
+    MICROLIBRARY_BUILD_INTERACTIVE_TESTS
+    "microlibrary: Build interactive tests."
+    OFF
+    )
+
+# general configuration validation
+
+set(
+    MICROLIBRARY_HIL_LIST
+    "ALL"
+    "CUSTOM"
+    )
+if( NOT MICROLIBRARY_HIL IN_LIST MICROLIBRARY_HIL_LIST )
+    message(
+        FATAL_ERROR
+        "'${MICROLIBRARY_HIL}' is not a valid value for MICROLIBRARY_HIL.\n"
+        "Please set it to a value from the first column of the following list:\n"
+        "    ALL        All officially supported HILs.\n"
+        "    CUSTOM     A HIL that is not officially supported.\n"
+        )
+endif( NOT MICROLIBRARY_HIL IN_LIST MICROLIBRARY_HIL_LIST )
+
+set(
+    MICROLIBRARY_TARGET_LIST
+    "DEVELOPMENT_ENVIRONMENT"
+    "HARDWARE"
+    )
+if( NOT MICROLIBRARY_TARGET IN_LIST MICROLIBRARY_TARGET_LIST )
+    message(
+        FATAL_ERROR
+        "'${MICROLIBRARY_TARGET}' is not a valid value for MICROLIBRARY_TARGET.\n"
+        "Please set it to a value from the first column of the following list:\n"
+        "    DEVELOPMENT_ENVIRONMENT    Development environment.\n"
+        "    HARDWARE                   Hardware.\n"
+        )
+endif( NOT MICROLIBRARY_TARGET IN_LIST MICROLIBRARY_TARGET_LIST )
+
+if( MICROLIBRARY_HIL STREQUAL "ALL" AND NOT MICROLIBRARY_TARGET STREQUAL "DEVELOPMENT_ENVIRONMENT" )
+    message(
+        FATAL_ERROR
+        "MICROLIBRARY_TARGET must be 'DEVELOPMENT_ENVIRONMENT' if MICROLIBRARY_HIL is 'ALL'"
+        )
+endif( MICROLIBRARY_HIL STREQUAL "ALL" AND NOT MICROLIBRARY_TARGET STREQUAL "DEVELOPMENT_ENVIRONMENT" )
+
+if( MICROLIBRARY_HIL STREQUAL "CUSTOM" AND CMAKE_SOURCE_DIR STREQUAL "${PROJECT_SOURCE_DIR}" )
+    message(
+        FATAL_ERROR
+        "Standalone builds are not supported if MICROLIBRARY_HIL is 'CUSTOM'"
+        )
+endif( MICROLIBRARY_HIL STREQUAL "CUSTOM" AND CMAKE_SOURCE_DIR STREQUAL "${PROJECT_SOURCE_DIR}" )
```

The quoting and explicit variable de-referencing should not be necessary according to the documentation.
```suggestion
if( MICROLIBRARY_HIL STREQUAL "CUSTOM" AND CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR )
    message(
        FATAL_ERROR
        "Standalone builds are not supported if MICROLIBRARY_HIL is 'CUSTOM'"
        )
endif( MICROLIBRARY_HIL STREQUAL "CUSTOM" AND CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR )
```